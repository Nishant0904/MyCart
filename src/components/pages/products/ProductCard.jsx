import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import starIcon from '../../../assets/icons/star.png'
import Icon from '../../reusables/Icon'
import ProductQuantitySelector from './ProductQuantitySelector'
import formatToUSD from '../../../utils/formatToUSD'

const ProductCard = ({ product, handleAddToCart }) => {
    const productPagePath = `/products/${product.id}`

    return (
        <StyledCard data-testid='product-card'>
            <CardImageLink to={productPagePath} data-testid='product-card__image-link'>
                <img
                    src={product.image}
                    alt={product.title}
                    title={product.title}
                />
            </CardImageLink>
            <CardInfoContainer>
                <CardInfoSection>
                    <Link to={productPagePath} data-testid='product-card__title-link'>
                        <h4>{product.title}</h4>
                    </Link>
                    <CardInfoRating>
                        <Icon src={starIcon} alt='star icon' />
                        <p>{product.rating.rate}</p>
                    </CardInfoRating>
                </CardInfoSection>
                <p>{formatToUSD(product.price)}</p>
                <ProductQuantitySelector
                    product={product}
                    handleAddToCart={handleAddToCart}
                />
            </CardInfoContainer>
        </StyledCard>
    )
}

const StyledCard = styled.div`
    display: flex;
    flex-direction: column;
    gap: 8px;
    padding: 24px;
    max-width: 336px;
`

const CardImageLink = styled(Link)`
    align-self: center;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-grow: 1;

    img {
        width: 100%;
    }
`

const CardInfoContainer = styled.div`
    display: flex;
    flex-direction: column;
    gap: 8px;

    h4 {
        font-size: 1rem;
        font-weight: 500;
    }
`

const CardInfoSection = styled.div`
    display: flex;
    justify-content: space-between;
    align-items: start;
    gap: 8px;
`

const CardInfoRating = styled.div`
    display: flex;
    align-items: center;
    flex-wrap: nowrap;
    gap: 2px;
`

ProductCard.propTypes = {
    product: PropTypes.object.isRequired,
    handleAddToCart: PropTypes.func.isRequired,
}

export default ProductCard
