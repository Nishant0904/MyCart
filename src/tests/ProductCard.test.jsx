import { describe, it, expect } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import ProductCard from '../components/pages/products/ProductCard'
import { BrowserRouter } from 'react-router-dom'
import userEvent from '@testing-library/user-event'

const mockProduct = {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
        'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
        rate: 3.9,
        count: 120,
    },
}

describe('Product Card test', () => {
    it('Product card renders correctly', () => {
        render(
            <ProductCard product={mockProduct} handleAddToCart={() => {}} />,
            { wrapper: BrowserRouter },
        )
        expect(screen.getByTestId('product-card')).toBeInTheDocument()
        expect(screen.getByTestId('product-card')).toMatchSnapshot()
    })

    it('Product card title link takes user to single product page', async () => {
        const user = userEvent.setup()

        render(
            <ProductCard product={mockProduct} handleAddToCart={() => {}} />,
            { wrapper: BrowserRouter },
        )

        const productTitleLink = screen.getByTestId('product-card__title-link')
        await user.click(productTitleLink)

        await waitFor(() => {
            expect(window.location.pathname).toBe(`/products/${mockProduct.id}`)
        })
    })

    it('Product card image link takes user to single product page', async () => {
        const user = userEvent.setup()

        render(
            <ProductCard product={mockProduct} handleAddToCart={() => {}} />,
            { wrapper: BrowserRouter },
        )

        const productImageLink = screen.getByTestId('product-card__image-link')
        await user.click(productImageLink)

        await waitFor(() => {
            expect(window.location.pathname).toBe(`/products/${mockProduct.id}`)
        })
    })
})
