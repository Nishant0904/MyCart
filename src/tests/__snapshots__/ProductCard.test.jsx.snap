// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Card test > Product card renders correctly 1`] = `
<div
  class="sc-eDLKkx ceMjzl"
  data-testid="product-card"
>
  <a
    class="sc-jTQCzO imamgb"
    data-testid="product-card__image-link"
    href="/products/1"
  >
    <img
      alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
      src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      title="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
    />
  </a>
  <div
    class="sc-gLLuof kTfTWK"
  >
    <div
      class="sc-iBdnpw jphUIh"
    >
      <a
        data-testid="product-card__title-link"
        href="/products/1"
      >
        <h4>
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </h4>
      </a>
      <div
        class="sc-fsYfdN cwoSHM"
      >
        <img
          alt="star icon"
          class="sc-beySPh htjDQV"
          src="/src/assets/icons/star.png"
        />
        <p>
          3.9
        </p>
      </div>
    </div>
    <p>
      $109.95
    </p>
    <div
      class="sc-dmyCSP gUwTfa"
      data-testid="product-quantity-selector"
    >
      <div
        class="sc-hLQSwg tApZJ"
      >
        <button
          class="sc-guDLey hYdscr"
          data-testid="product-quantity__decrement-btn"
        >
          â€“
        </button>
        <input
          data-testid="product-quantity__number-input"
          type="number"
          value="1"
        />
        <button
          class="sc-guDLey hYdscr"
          data-testid="product-quantity__increment-btn"
        >
          +
        </button>
      </div>
      <button
        class="sc-guDLey hYdscr"
        data-testid="product-quantity__add-to-cart-btn"
      >
        Add to Cart
      </button>
    </div>
  </div>
</div>
`;
